### PR TITLE
[delimitedtext] Fix datetime,date field type ignored (in CSVT et cie)

### LINKED ISSUE #20882: Import CSV via "add delimited text layer" ignores Date fields in CSVT
Author Name: **Jeremy Kidwell** (Jeremy Kidwell)
Original Redmine Issue: [12783](https://issues.qgis.org/issues/12783)
Affected QGIS version: master
Redmine category:data_provider/delimited_text_

---
When importing a CSV file with fields defined in CSVT, QGIS ignores fields designated as "Date" and imports them as QSTRING / text. I have confirmed that my CSVT file is properly formatted and the import does make use of field designations for "Real","Integer","String".

Data samples as follows:

GIS-Locations_2.4.csvt:
```"Real","Real","Integer","String","String","String","String","String","String","String","Integer","Real","Date","Date","Date","Date","String","String","String","String","String"
```

GIS-Locations_2.4.csv (first few lines):
```"X","Y","accuracy code","postcode","unique-id","name","network","denomination","notes","website","awards","icon-size2","award1","award2","award3","award4","wildlife-biodiversity","solar","date-built","churches trust URL","OS ref"
"349344.8977","1053019.497","2","KW17 2BU","2203","St. Ann's, Papa Westray","Orkney Islands","C of S","","","4","3","2003-10-09","2006-08-19","2009-07-07","2013-11-01","","x","1841","http://www.scotlandschurchestrust.org.uk/church/st-anns-papa-westray","HY495516"
"345691","1046915","3","KW17 2DB","2202","Westray Church","Orkney Islands","C of S","","","4","3","2003-10-09","2006-08-19","2009-07-08","2013-11-01","","","1846","http://www.scotlandschurchestrust.org.uk/church/westray-parish-kirk","HY457462"
"272400.702","655262.3425","2","ML3 7DT","2801","St John's Church Of Scotland, Hamilton","South Lanarkshire","C of S","","http://www.stjohnshamilton.org.uk","3","2.5","2003-05-01","2006-03-05","2009-07-09","","","","1835","http://www.scotlandschurchestrust.org.uk/church/st-johns-church-hamilton","NS724553"
```

I have confirmed that running GDAL from the command line gives the expected result (using "ogrinfo GIS-Locations_2.4.csv -al").

I initially thought this might be user error, and so put up a post on gis.stackexchange.com (http://gis.stackexchange.com/questions/146291/importing-csv-file-with-csvt-data-in-qgis-2-8-will-not-assign-type-date-to-f) but have concluded (with help) that this is likely a bug in the current build of QGIS. I'm running 2.8.2 (but also had problem in 2.8.1) on MacOS 10.10.3 and have also confirmed that this is an issue in QGIS windows (running inside virtualbox on Win XP).

---

- [table_in_spatialite.PNG](https://issues.qgis.org/attachments/download/8721/table_in_spatialite.PNG) (Saber Razmjooei)

---

Related issue(s): #23736 (relates), #26676 (relates)
Redmine related issue(s): [15816](https://issues.qgis.org/issues/15816), [18789](https://issues.qgis.org/issues/18789)

---

### PR DESCRIPTION
**Pull request by nirvn**
## Description

This PR insures that the QGIS delimited text provider handles datetime, date, and time field types properly (by respecting the type declared in a .csvt or auto-detecting type). It fixes #20882 ( which was also raised in stackexchange here: https://gis.stackexchange.com/questions/364292/qgis3-import-csv-file-with-datetime-values/364295 ).

It also harmonizes the handling of CSV files with GDAL, which is important as users often get confused when the delimited text provider vs ogr provider behave differently.

### CODE REVIEW

File: src/providers/delimitedtext/qgsdelimitedtextprovider.cpp [hunk 1]
```diff
@@ -604,25 +619,42 @@ void QgsDelimitedTextProvider::scanFile( bool buildIndexes )
...
         {
           value.replace( mDecimalPoint, QLatin1String( "." ) );
         }
         ( void )value.toDouble( &couldBeDouble[i] );
       }
+
+      if ( couldBeDateTime[i] )
+      {
+        QDateTime dt = QDateTime::fromString( value, Qt::ISODate );
+        couldBeDateTime[i] = ( dt.isValid() && value.length() > 10 );
```
Comment (nyalldawson) [inline src/providers/delimitedtext/qgsdelimitedtextprovider.cpp:639]: i'd suggest moving the value.length() check to happen before the attempted conversion -- you could shortcut a bunch of work this way

File: tests/src/python/test_qgsdelimitedtextprovider.py [hunk 2]
```diff
@@ -907,6 +907,25 @@ def test_046_M(self):
...
+        url.addQueryItem("xField", "X")
+        url.addQueryItem("yField", "Y")
+        url.addQueryItem("spatialIndex", "no")
+        url.addQueryItem("subsetIndex", "no")
+        url.addQueryItem("watchFile", "no")
+
+        vl = QgsVectorLayer(url.toString(), 'test', 'delimitedtext')
+        assert vl.isValid(), "{} is invalid".format(basetestfile)
+        assert vl.fields().at(4).type() == QVariant.DateTime
+        assert vl.fields().at(5).type() == QVariant.Date
```
Comment (nyalldawson) [inline tests/src/python/test_qgsdelimitedtextprovider.py:927]: why not include a time check here too?

**Comment by nyalldawson** [comment]
Looking good!

**Comment by nirvn** [comment]
Here's a good example of ogr vs delimited text provider confusion and unmet expectation: https://gis.stackexchange.com/questions/146291/importing-csv-file-with-csvt-data-in-qgis-2-8-will-not-assign-type-date-to-f

**Pull request merged**